### PR TITLE
add proposal for robot permission expansion

### PR DESCRIPTION
**Pull request by wy65701436**

### CODE REVIEW

**Review by chlins — APPROVED** [review]
lgtm

File: proposals/new/Robot-Account-Expand.md [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+This section lists the user stories for the different personas interacting with robot account.
+
+* Personas
+
+Robot Account is a System Administrator and Project Administrator operation in Harbor.
+
+* User Stories
+
+1. As a system administrator, I can enable or disable the prohibited scope set via a configuration option.
```
Comment (reasonerjt) [inline proposals/new/Robot-Account-Expand.md:35]: Do we really need to introduce the concept "prohibited scope" to user or developer?  The prohibited scope is not more sensitive than the existing ones.

File: proposals/new/Robot-Account-Expand.md [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+4. As a system administrator, I can identify the creator of each robot account by performing an SQL query in the database.
+5. A robot account can create another robot account, but the new account’s scope must be less than or equal to that of the creator.
+6. A robot account created by another robot can only be updated or deleted by either the human with the relevant permissions, the creator of the robot account, or the robot account itself.
+
+## Scheme Change
+
+Add a new column of creator for table robot.
+
+```
+ALTER TABLE robot ADD COLUMN IF NOT EXISTS creator varchar(255);
```
Comment (reasonerjt) [inline proposals/new/Robot-Account-Expand.md:47]: IMO more details are needed for the value of this column, how to differentiate a user from a robot account?

File: proposals/new/Robot-Account-Expand.md [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+|   Robot   |    Create    |   Y    |
+|   Robot   | Delete(self) |   Y    |
+
+## UI
+
+![expenaded_permissons](../images/robot-expand-permission/robot1.png)
+
+![audit_log](../images/robot-expand-permission/robot2.png)
+
+## Data provider
```
Comment (reasonerjt) [inline proposals/new/Robot-Account-Expand.md:101]: Will we need to make update to the security context of a robot to list the permissions?  Otherwise how do we ensure there's no elevation?

File: proposals/new/Robot-Account-Expand.md [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+Robot Account is a System Administrator and Project Administrator operation in Harbor.
+
+* User Stories
+
+1. As a system administrator, I can enable or disable the prohibited scope set via a configuration option.
+2. As a system/project administrator, I can create a project-level robot account with the selected access scope, including the prohibited scope set.
+3. The creation and deletion of robot accounts will be recorded in the audit log.
+4. As a system administrator, I can identify the creator of each robot account by performing an SQL query in the database.
+5. A robot account can create another robot account, but the new account’s scope must be less than or equal to that of the creator.
+6. A robot account created by another robot can only be updated or deleted by either the human with the relevant permissions, the creator of the robot account, or the robot account itself.
```
Comment (reasonerjt) [inline proposals/new/Robot-Account-Expand.md:40]: I wish to suggest in addition to the items in user story, we should zoom into the scenario where robot account managing robot account, b/c this is where the most requirement come from and there are quite a few technical decisions we made that should be documented for discussion with other maintainers and for reference in future

File: proposals/new/Robot-Account-Expand.md [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+
+Robot Account is a System Administrator and Project Administrator operation in Harbor.
+
+* User Stories
+
+1. As a system administrator, I can enable or disable the prohibited scope set via a configuration option.
+2. As a system/project administrator, I can create a project-level robot account with the selected access scope, including the prohibited scope set.
+3. The creation and deletion of robot accounts will be recorded in the audit log.
+4. As a system administrator, I can identify the creator of each robot account by performing an SQL query in the database.
+5. A robot account can create another robot account, but the new account’s scope must be less than or equal to that of the creator.
```
Comment (reasonerjt) [inline proposals/new/Robot-Account-Expand.md:39]: Does it apply to both system-level and project-level?

File: proposals/new/Robot-Account-Expand.md [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+|   ExportCVE   |    Create     |   Y    |
+|   LdapUser    |     List      |   Y    |
+|   LdapUser    |    Create     |   Y    |
+|  User-Group   |     List      |   Y    |
+|  User-Group   |    Create     |   Y    |
+|  User-Group   |     Read      |   Y    |
+|  User-Group   |    Update     |   Y    |
+|  User-Group   |    Delete     |   Y    |
+|     Robot     |     Read      |   Y    |
+|     Robot     | Update(self)  |   Y    |
```
Comment (reasonerjt) [inline proposals/new/Robot-Account-Expand.md:69]: This is not consistent with the item in "User Stories"
```
A robot account created by another robot can only be updated or deleted by either the human with the relevant permissions, the creator of the robot account, or the robot account itself.
```

File: proposals/new/accessory.md [hunk 7]
```diff
@@ -89,6 +89,7 @@ CREATE TABLE artifact_accessory (
 ### Accessory Interface
```
Comment (reasonerjt) [inline proposals/new/accessory.md:89]: How is this file relevant to this design?

File: proposals/new/Robot-Account-Expand.md (hunk 1 above)
Comment (wy65701436) [reply proposals/new/Robot-Account-Expand.md:35]: the story is not needed.

File: proposals/new/Robot-Account-Expand.md (hunk 2 above)
Comment (wy65701436) [reply proposals/new/Robot-Account-Expand.md:47]: Since the creator is identified by the full name of either a Harbor user or a robot account, and robot account names include a unique prefix, it allows easy distinction between robot accounts and human users.

File: proposals/new/Robot-Account-Expand.md (hunk 5 above)
Comment (wy65701436) [reply proposals/new/Robot-Account-Expand.md:39]: yes

File: proposals/new/Robot-Account-Expand.md (hunk 6 above)
Comment (wy65701436) [reply proposals/new/Robot-Account-Expand.md:69]: updated

File: proposals/new/accessory.md (hunk 7 above)
Comment (wy65701436) [reply proposals/new/accessory.md:89]: removed.

File: proposals/new/Robot-Account-Expand.md (hunk 3 above)
Comment (wy65701436) [reply proposals/new/Robot-Account-Expand.md:101]: The API handler knows that the current user is a robot. With this knowledge, it can determine both the scope of the creator and the scope of the robot. The handler then decides whether to proceed based on the comparison of these scopes.

File: proposals/new/Robot-Account-Expand.md (hunk 4 above)
Comment (wy65701436) [reply proposals/new/Robot-Account-Expand.md:40]: I will add a separate section for the robot that created by another robot.

File: proposals/new/Robot-Account-Expand.md [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+1. Is there any limitation for the number of robot accounts that can be created by an existing robot account?
+2. Is there any limitation for the number of robot accounts you can create within a single project?
+
+## Scheme Change
+
+Add a new column of creator for table robot.
+
+```
+ALTER TABLE robot ADD COLUMN IF NOT EXISTS creator varchar(255);
+UPDATE robot SET creator = 'unknown' WHERE creator IS NULL;
```
Comment (reasonerjt) [inline proposals/new/Robot-Account-Expand.md:78]: Is there anything to prevent a user from being registered as "unknown"?

File: proposals/new/Robot-Account-Expand.md [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+
+1. the creator name of a robot account is a good way to capture the contact person for the account. For new robots that created after this proposal be introduced, Harbor will populate the name
+of the creator when creating the robot account. 
+2. The new audit logs for robot account creation and deletion are good way for you to track the lifecycle of the created robots.
+
+![audit_log](../images/robot-expand-permission/robot2.png)
+
+## Robot accounts that are created and managed by another robot accounts
+
+Robot accounts are principals, it means that you can grant service account to Harbor resources. However, from the perspective of preventing privilege escalation, generally, a service account
```
Comment (reasonerjt) [inline proposals/new/Robot-Account-Expand.md:53]: Is "service account" equivalent to "robot account"?  If the answer is yes, why not unify them?

File: proposals/new/Robot-Account-Expand.md [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+Robot accounts are principals, it means that you can grant service account to Harbor resources. However, from the perspective of preventing privilege escalation, generally, a service account 
+cannot grant roles that are higher or more powerful than the roles it possesses. When a robot account creates a new service account, the new service account doesn't automatically inherit any roles
+or permissions.
+
+Creation: If a robot account has been given the role of robot account creation, it can create or manage other service accounts but only within the permissions of its own roles.
+1. any project level robot account can be created by a system or project level robot account who with the robot creation permission.
+2. any system level robot account can be created by a system level robot account who with the robot creation permission.
+
+Deletion/Update: A service account cannot assign permissions to another service account that exceed its own permissions.
+1. any robot account that created by another robot can be updated by the creator robot or the itself.
```
Comment (reasonerjt) [inline proposals/new/Robot-Account-Expand.md:62]: By "the itself" do you mean "itself"?  
Are you saying a robot account can remove itself even if it doesn't have permission to delete "robot account"?  
How does it work before we allow a robot account to create a robot account?

File: proposals/new/Robot-Account-Expand.md [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+
+Robot Account is a System Administrator and Project Administrator operation in Harbor.
+
+* User Stories
+
+1. As a system/project administrator, I can create a project-level robot account with the selected access scope, including the banned scope set in the harbor v2.10.0.
+2. The creation and deletion of robot accounts will be recorded in the audit log.
+3. As a system administrator, I can identify the creator of each robot account by performing an SQL query in the database.
+4. A robot account can create another robot account, but the new account’s scope must be less than or equal to that of the creator.
+5. A robot account created by another robot can only be updated or deleted by either the human with the relevant permissions, the creator of the robot account, or the robot account itself.
```
Comment (reasonerjt) [inline proposals/new/Robot-Account-Expand.md:39]: 1. When a robot account updates a robot account that it created, we should also make sure there's elevation.
2. I don't quite understand why we wanna limit the deletion.  If a system-level robot account has the permission to delete a robot account, why we don't allow it to delete any other robot account?  Isn't this model a little easier to understand?

File: proposals/new/Robot-Account-Expand.md (hunk 9 above)
Comment (wy65701436) [reply proposals/new/Robot-Account-Expand.md:53]: sure, I will rephrase it.

File: proposals/new/Robot-Account-Expand.md (hunk 10 above)
Comment (wy65701436) [reply proposals/new/Robot-Account-Expand.md:62]: yes, I mean only when the creator has the permission of robot account deletion can delete the created robot.

As for the robot that created by another robot,
1. any human with the correct permission can remove it.
2. if the creator has the robot deletion permission, it can remove the created robots.
3. if itself was granted the robot deletion permission when it was created, it can remove itself.

File: proposals/new/Robot-Account-Expand.md (hunk 11 above)
Comment (wy65701436) [reply proposals/new/Robot-Account-Expand.md:39]: 1, limitation of updating is the same as creation. 
2, as for the system level robot, it can remove any other robots if it has the relevant permission.

File: proposals/new/Robot-Account-Expand.md [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+Robot accounts are principals, it means that you can grant robot account to Harbor resources. However, from the perspective of preventing privilege escalation, generally, a robot account 
+cannot grant roles that are higher or more powerful than the roles it possesses. When a robot account creates a new robot account, the new robot account doesn't automatically inherit any roles
+or permissions.
+
+Creation: If a robot account has been given the role of robot account creation, it can create or manage other robot accounts but only within the permissions of its own roles.
+1. any project level robot account can be created by a system or project level robot account who with the robot creation permission.
+2. any system level robot account can be created by a system level robot account who with the robot creation permission.
+
+Deletion/Update: A robot account cannot assign permissions to another robot account that exceed its own permissions.
+1. any robot account that created by another robot can be updated by the creator robot or the itself, provided that the update permission has been granted to the creator or the robot itself.
```
Comment (stonezdj) [inline proposals/new/Robot-Account-Expand.md:62]: `the itself` should be `itself`

File: proposals/new/Robot-Account-Expand.md [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+1. the creator name of a robot account is a good way to capture the contact person for the account. For new robots that created after this proposal be introduced, Harbor will populate the name
+of the creator when creating the robot account. 
+2. The new audit logs for robot account creation and deletion are good way for you to track the lifecycle of the created robots.
+
+![audit_log](../images/robot-expand-permission/robot2.png)
+
+## Robot accounts that are created and managed by another robot accounts
+
+Robot accounts are principals, it means that you can grant robot account to Harbor resources. However, from the perspective of preventing privilege escalation, generally, a robot account 
+cannot grant roles that are higher or more powerful than the roles it possesses. When a robot account creates a new robot account, the new robot account doesn't automatically inherit any roles
```
Comment (stonezdj) [inline proposals/new/Robot-Account-Expand.md:54]: replace role with permission, avoid using role because the role is only used in project member,

**Review by stonezdj — APPROVED** [review]

**Review by MinerYang — APPROVED** [review]
lgtm

**Review by reasonerjt — APPROVED** [review]

File: proposals/new/Robot-Account-Expand.md [hunk 14]
```diff
@@ -0,0 +1,194 @@
...
+1. As a system/project administrator, I can create a system-level/project-level robot account with the selected access scope, including the banned scope set in the harbor v2.10.0.
+2. The creation and deletion of robot accounts will be recorded in the audit log.
+3. As a system administrator, I can identify the creator of each robot account by performing an SQL query in the database(There is no API to directly handle this).
+4. A robot account can create another robot account, but the nested robot account’s scope must be less than or equal to that of the creator robot.
+
+## Keep track of robot accounts
+
+Over time, as you create more and more robot accounts, and you may lose track of who creates the robot account and which robot account is used for.
+
+1. the creator information of a robot account is a good way to capture the contact person for the account. For new robots that created after this proposal be introduced, Harbor will populate the type and reference
```
Comment (Vad1mo) [inline proposals/new/Robot-Account-Expand.md:48]: This statement is difficult to understand, can you break down and elaborate a bit more.


What is the _"creator information"_? > Do you maybe mean the actual entity that created the robot account? Human who created the robot account or robot account create a robot account. 

> Harbor will populate the type and reference.

What type of reference? Reference to what or whom.

File: proposals/new/Robot-Account-Expand.md [hunk 15]
```diff
@@ -0,0 +1,194 @@
...
+of the creator when creating the robot account. 
+2. The new audit logs for robot account creation and deletion are good way for you to track the lifecycle of the created robots.
```
Comment (Vad1mo) [inline proposals/new/Robot-Account-Expand.md:50]: We have already audit logs that are capable of tracking that information, we just don't use them for logging who created roboto accounts. Is it needed to extend the audit logs with new fields?

File: proposals/new/Robot-Account-Expand.md [hunk 16]
```diff
@@ -0,0 +1,194 @@
...
+
+**Note**: Since the creator robot account’s permissions can be updated without impacting its nested accounts, this can lead to situations where the **nested robot account has more powerful permissions** than its creator.
+
+Deletion: 
+1. any nested robot account can be deleted by a robot with the appropriate robot deletion permissions.
+
+When a creator robot account is removed, the nested robot accounts it created are **not automatically removed**. These nested accounts will continue to exist and function independently, as long as they retain the necessary permissions.
+
+Open questions:
+1. Is there a limitation on the number of robot accounts that can be created by a creator robot account?
```
Comment (Vad1mo) [inline proposals/new/Robot-Account-Expand.md:78]: I would say no, why should we restrict it and what should be the right number 100 or 10000?

File: proposals/new/Robot-Account-Expand.md [hunk 17]
```diff
@@ -0,0 +1,194 @@
...
+2. Is there a limitation on the number of robot accounts that can be created within a single project?
```
Comment (Vad1mo) [inline proposals/new/Robot-Account-Expand.md:79]: no, need to make artificial restrictions here too.

File: proposals/new/Robot-Account-Expand.md [hunk 18]
```diff
@@ -0,0 +1,194 @@
...
+Originally, the Harbor UI offered 16 permission options for users, but this was expanded to 40 in v2.10. To address security concerns, Harbor currently restricts the creation of robots with permissions related to user (normal and robot), group, and quota creation/deletion. However, this restriction has disrupted some users' existing pipelines that rely on robots with these prohibited permissions.
+
+## Solution
+
+This proposal suggests introducing a permission enhancement that allows admins to create robot accounts with a restricted permission set—introduced in v2.10—by specifying the selected permissions.
+Additionally, by recording the creator of each robot account in the database and audit logs, this feature will help system administrators mitigate security concerns and ensure accountability.
+
+## Non Goal
+
+1. No support for configuring the banned permissions in the harbor v2.10.0
```
[the remark this answers is not in the document]
Comment (wy65701436) [reply proposals/new/Robot-Account-Expand.md:22]: the permission in the data grid, like the CRUD of user.

File: proposals/new/Robot-Account-Expand.md (hunk 14 above)
Comment (wy65701436) [reply proposals/new/Robot-Account-Expand.md:48]: sure, I will add more details for these two added columns.

the creator type is pointing whether the creator is a human or robot.
the reference is the reference to the table of harbor user or robot, actually it is the ID of user of robot.

File: proposals/new/Robot-Account-Expand.md (hunk 16 above)
Comment (wy65701436) [reply proposals/new/Robot-Account-Expand.md:78]: I just wanna point this out, since we introduce the ability for robot to create another robot. And to avoid abusing, should we set some rules for robots?

File: proposals/new/Robot-Account-Expand.md (hunk 17 above)
Comment (wy65701436) [reply proposals/new/Robot-Account-Expand.md:79]: ditto

File: proposals/new/Robot-Account-Expand.md (hunk 15 above)
Comment (wy65701436) [reply proposals/new/Robot-Account-Expand.md:50]: there is no new field will be introduced, but just records the creation and deletion of a robot account in the audit, and if the admin wanna know the creator, then he/she can get it from the data base.
